lib/notifications/display.dart

```diff
@@ -39,6 +57,110 @@ class NotificationChannelManager {
   @visibleForTesting
   static final kVibrationPattern = Int64List.fromList([0, 125, 100, 450]);
 
+  /// Generates an Android resource URL for the given resource name and type.
+  ///
+  /// For example, for a resource `@raw/chime3`, where `raw` would be the
+  /// resource type and `chime3` would be the resource name it generates the
+  /// following URL:
+  ///   `android.resource://com.zulip.flutter/raw/chime3`
+  ///
+  /// Based on: https://stackoverflow.com/a/38340580
+  static Uri _resourceUrlFromName({
+    required String resourceTypeName,
+    required String resourceEntryName,
+  }) {
+    const packageName = 'com.zulip.flutter'; // TODO(#407)
+
+    // URL scheme for Android resource url.
+    // See: https://developer.android.com/reference/android/content/ContentResolver#SCHEME_ANDROID_RESOURCE
+    const schemeAndroidResource = 'android.resource';
+
+    return Uri(
+      scheme: schemeAndroidResource,
+      host: packageName,
+      pathSegments: <String>[resourceTypeName, resourceEntryName],
+    );
+  }
+
+  /// Prepare our notification sounds; return a URL for our default sound.
+  ///
+  /// Where possible, this copies each of our notification sounds into shared storage
+  /// so that the user can choose between them in the system notification settings.
+  ///
+  /// Returns a URL for our default notification sound: either in shared storage
+  /// if we successfully copied it there, or else as our internal resource file.
+  static Future<String> _ensureInitNotificationSounds() async {
+    String defaultSoundUrl = _resourceUrlFromName(
+      resourceTypeName: 'raw',
+      resourceEntryName: kDefaultNotificationSound.resourceName).toString();
+
+    final shouldUseResourceFile = switch (await ZulipBinding.instance.deviceInfo) {
+      // Before Android 10 Q, we don't attempt to put the sounds in shared media storage.
+      // Just use the resource file directly.
+      // TODO(android-sdk-29): Simplify this away.
+      AndroidDeviceInfo(:var sdkInt) => sdkInt <= 28,
```

Comment:
nit:
```suggestion
      // Before Android 10 Q, we don't attempt to put the sounds in shared media storage.
      // Just use the resource file directly.
      // TODO(android-sdk-29): Simplify this away.
      AndroidDeviceInfo(:var sdkInt) => sdkInt < 29,
```

In general I think it's cleanest to express version conditions as a single threshold, which is the first version where the change appeared: so conditions look like "actual >= threshold" or "actual < threshold", but never "actual <= one-less-than-threshold".

That's because if things are instead expressed with two neighboring numbers (like 28 and 29 here), it starts getting harder to keep straight in one's head. I noticed this in the previous round, but was prompted to comment on it by my comment just before this one :slightly_smiling_face: